SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -76,42 +92,111 @@ void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_Compo
 	}
 }
 
+void SpatialVirtualWorkerTranslationManager::SetKnownServerSystemEntities(TArray<Worker_EntityId> ServerSystemEntities)
+{
+	KnownServerSystemEntities = ServerSystemEntities;
+}
+
+void SpatialVirtualWorkerTranslationManager::OnWorkerDisconnected(const Worker_EntityId DisconnectedSystemEntityId)
+{
+	// If we finish this loop without hitting anything, we assume it's a client that disconnected.
+	for (auto& VirtualWorkerInfo : VirtualToPhysicalWorkerMapping)
+	{
+		if (VirtualWorkerInfo.Value.ServerSystemWorkerEntity == DisconnectedSystemEntityId)
+		{
+			UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT("VTM identified disconnected server %s"),
+				   *VirtualWorkerInfo.Value.PhysicalWorkerName);
+			UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT(" - virtual worker %ld"), VirtualWorkerInfo.Key);
+			UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT(" - partition %lld"), VirtualWorkerInfo.Value.PartitionId);
+			UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT(" - system entity %lld"),
+				   VirtualWorkerInfo.Value.ServerSystemWorkerEntity);
+
+			ReclaimCrashedVirtualWorker(VirtualWorkerInfo.Key);
+		}
+	}
+}
+
+void SpatialVirtualWorkerTranslationManager::CleanupTranslatorMappingAfterAuthorityChange()
+{
+	TArray<VirtualWorkerId> CrashedVirtualWorkers;
+
+	for (auto& VirtualWorkerInfo : VirtualToPhysicalWorkerMapping)
+	{
+		if (!KnownServerSystemEntities.Contains(VirtualWorkerInfo.Value.ServerSystemWorkerEntity))
+		{
+			UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning, TEXT(" - detected crashed virtual worker %ld"),
+				   VirtualWorkerInfo.Key);
+			CrashedVirtualWorkers.Add(VirtualWorkerInfo.Key);
+		}
+	}
+
+	// In a different loop in case multiple servers crashed that we're trying to clean up, and we don't want to mess with
```

Comment:
and we don't want to mess with VirtualToPhysicalWorkerMapping ?

Reply:
Sorry, updated locally:
```
	// This is a different loop in case multiple partitions are now missing auth server system entities, and we don't
	// want to mess with deleting from collections while iterating through them.
```